lib/winapi/sync.c

```diff
@@ -1,11 +1,145 @@
 #include <synchapi.h>
 #include <assert.h>
+#include <stdatomic.h>
 #include <stdbool.h>
 #include <processthreadsapi.h>
 #include <winbase.h>
 #include <winerror.h>
 #include <xboxkrnl/xboxkrnl.h>
 
+static BOOL WINAPI condvar_init (PINIT_ONCE InitOnce, PVOID Parameter, PVOID *Context)
+{
+    PCONDITION_VARIABLE ConditionVariable = Parameter;
+    NTSTATUS status;
+    status = NtCreateEvent(&ConditionVariable->eventHandles[0], NULL, SynchronizationEvent, FALSE);
+    assert(NT_SUCCESS(status));
+    status = NtCreateEvent(&ConditionVariable->eventHandles[1], NULL, NotificationEvent, FALSE);
+    assert(NT_SUCCESS(status));
+    return true;
+}
+
+VOID InitializeConditionVariable (PCONDITION_VARIABLE ConditionVariable)
+{
+    ConditionVariable->initOnce.Ptr = 0;
+    ConditionVariable->eventHandles[0] = INVALID_HANDLE_VALUE;
+    ConditionVariable->eventHandles[1] = INVALID_HANDLE_VALUE;
+    atomic_store(&ConditionVariable->waitCount, 0);
+
+    InitOnceExecuteOnce(&ConditionVariable->initOnce, condvar_init, ConditionVariable, NULL);
```

Comment:
What if this function fails? Other instances, too.

Reply:
In that case this either deserves an `assert` or at very least a comment.
I have tried to follow the logic, and there appear to be various non-trivial ways in which `InitOnceExecuteOnce` could fail (through `InitOnceBeginInitialize` / `InitOnceComplete`).

A similar instance is the call to `InitOnceComplete` in `InitOnceExecuteOnce` where the return value is never checked. I assume that the assumption is that it always succeeds?
An `assert` could help to clarify if this is assumed, or a comment would help to understand why that wouldn't always be the case.

Such assertions would mostly help with self-checking and finding race-conditions or similar things. Given the complexity of the code I think it would be a good idea to add them.

Reply:
I've added the requested asserts now.